### PR TITLE
[Bugfix] Fix RMSNorm kernels to multiply in weight's native dtype

### LINKED ISSUE #42325: [Bug]: RMSNorm kernel ignores weight dtype, always uses FP32 (regression in v0.20.0)
### Your current environment

This is a kernel implementation bug affecting all platforms with CUDA. Environment details are not relevant as the issue is in the source code itself (`csrc/layernorm_kernels.cu`).

**Affected versions:** v0.20.0, v0.20.1, v0.20.2, and current main branch  
**Last working version:** v0.19.1

## Summary

vLLM's RMSNorm CUDA kernel (`csrc/layernorm_kernels.cu`) ignores the weight dtype and always multiplies in FP32, violating the Python reference specification in `vllm/ir/ops/layernorm.py`. This regression was introduced in v0.20.0 (commit 4d51588e23, PR #40860) and affects all subsequent versions including the current main branch.

## Actual Behavior (v0.20.0+)

The kernel casts weight to FP32 and performs multiplication in FP32:

```cpp
// csrc/layernorm_kernels.cu (lines 77-82, v0.20.0+)
#pragma unroll
for (int j = 0; j < VEC_SIZE; j++) {
  float x = static_cast<float>(src1.val[j]);
  float w = static_cast<float>(src2.val[j]);  // ❌ Always FP32
  dst.val[j] = static_cast<scalar_t>(x * s_variance * w);
}
```

## Expected Behavior (Specification)

The Python reference in `vllm/ir/ops/layernorm.py` specifies that multiplication should happen in **weight dtype**, not FP32:

```python
# vllm/ir/ops/layernorm.py (lines 19-21)
if weight is not None:
    x = x.to(weight.dtype) * weight  # ← Multiply in weight.dtype
return x.to(orig_dtype)
```

v0.19.1 and earlier correctly respected the weight dtype:

```cpp
// csrc/layernorm_kernels.cu (lines 77-82, v0.19.1)
#pragma unroll
for (int j = 0; j < VEC_SIZE; j++) {
  float x = static_cast<float>(src1.val[j]);
  dst.val[j] = ((scalar_t)(x * s_variance)) * src2.val[j];  // ✅ Respects weight dtype
}
```

## Version Timeline

| Version | Release Date | Status | Behavior |
|---------|--------------|--------|----------|
| v0.19.1 | April 15, 2026 | ✅ Correct | Respects weight dtype |
| v0.20.0 | April 27, 2026 | ❌ Bug | Always uses FP32 |
| v0.20.1 | May 4, 2026 | ❌ Bug | Always uses FP32 |
| v0.20.2 | May 5, 2026 | ❌ Bug | Always uses FP32 |
| main | Current | ❌ Bug | Always uses FP32 |

## Root Cause

The bug was introduced in commit 4d51588e23 (DeepSeek V4 feature PR #40860, April 26, 2026). This PR added FP8 quantization support and modified both the FP8 quantization kernels and the regular RMSNorm kernels.

**In the FP8 quantization context**, the FP32 multiplication is actually **correct** because the fused kernel must match the precision of the unfused composite operation (RMSNorm → BF16 cast → FP8 quant).

**However**, the same pattern was applied to the **regular RMSNorm kernel** where it's **incorrect** because it violates the specification that multiplication should happen in weight dtype.

This appears to be an unintended side effect rather than a deliberate design change.

### Precision Mismatch

When using BF16 weights (the common case for modern LLMs), the kernel:

- **Expected behavior**: Convert normalized values to BF16, multiply in BF16
- **Actual behavior**: Multiply in FP32, then convert to BF16

This causes measurable numerical divergence:

```
Max divergence from reference: ~0.03 to 0.06
```

### Cumulative Effects

For models with RMSNorm in multiple layers (e.g., attention Q/K normalization), the error compounds:

- Single layer: divergence = 0.03
- 20 layers: divergence = 124 (exponential growth)
- Final output match rate: ~2%

This can significantly affect model accuracy, especially for models that rely on precise normalization (e.g., LLaDA2 with Q/K normalization).

## Reproduction

```python
import torch
from vllm import _custom_ops as ops

def reference_rmsnorm(x, weight, eps=1e-6):
    """Reference matching vLLM IR ops spec."""
    orig_dtype = x.dtype
    x_fp32 = x.to(torch.float32)
    variance = x_fp32.pow(2).mean(-1, keepdim=True)
    x_normed = x_fp32 * torch.rsqrt(variance + eps)
    result = x_normed.to(weight.dtype) * weight  # Multiply in weight.dtype
    return result.to(orig_dtype)

# Test with BF16 weights
x = torch.randn(48, 128, dtype=torch.bfloat16, device='cuda')
weight = torch.randn(128, dtype=torch.bfloat16, device='cuda')

out = torch.empty_like(x)
ops.rms_norm(out, x, weight, 1e-6)

ref = reference_rmsnorm(x, weight)
diff = (out.float() - ref.float()).abs().max()

print(f"Max diff: {diff:.6e}")
# v0.19.1: ~0.0 (correct)
# v0.20.0+: ~0.03 (bug)
```

**Test results:**

| Version | Max Diff | Status |
|---------|----------|--------|
| v0.19.1 | ~0.0 | ✅ Matches specification |
| v0.20.0 | ~0.03 | ❌ Violates specification |
| v0.20.1 | ~0.03 | ❌ Violates specification |
| v0.20.2 | ~0.03 | ❌ Violates specification |

## Fix

Revert the regular RMSNorm kernel to v0.19.1 behavior:

```diff
// csrc/layernorm_kernels.cu (line ~77-82)
 #pragma unroll
 for (int j = 0; j < VEC_SIZE; j++) {
   float x = static_cast<float>(src1.val[j]);
-  float w = static_cast<float>(src2.val[j]);
-  dst.val[j] = static_cast<scalar_t>(x * s_variance * w);
+  dst.val[j] = ((scalar_t)(x * s_variance)) * src2.val[j];
 }
```

The same fix is needed in `fused_add_rms_norm_kernel` (line ~182):

```diff
 for (int idx = threadIdx.x; idx < hidden_size; idx += blockDim.x) {
   float x = (float)residual[blockIdx.x * hidden_size + idx];
-  float w = (float)weight[idx];
-  input[blockIdx.x * input_stride + idx] = (scalar_t)(x * s_variance * w);
+  input[blockIdx.x * input_stride + idx] = ((scalar_t)(x * s_variance)) * weight[idx];
 }
```

**Important**: Keep the FP32 cast in `layernorm_quant_kernels.cu` — it's correct there for FP8 quantization.

## Additional Context

- **Python reference specification**: [`vllm/ir/ops/layernorm.py`](https://github.com/vllm-project/vllm/blob/main/vllm/ir/ops/layernorm.py#L19-L21)
- **Introducing commit**: [4d51588e23](https://github.com/vllm-project/vllm/commit/4d51588e23) (DeepSeek V4 Rebased #40860, April 26, 2026)
- **Last good release**: v0.19.1 (April 15, 2026)
- **Python IR ops remain correct**: Only the CUDA kernel is affected; the Python reference in `vllm/ir/ops/layernorm.py` correctly follows the specification

This is a regression that breaks the contract between the IR ops specification and the kernel implementation. The fix restores specification compliance without affecting the FP8 quantization features added in v0.20.0.

---

### Before submitting a new issue...

- [x] I have searched for relevant issues, and asked the chatbot at the [documentation page](https://docs.vllm.ai/en/latest/)
- [x] I have tested this bug across multiple vLLM versions (v0.19.1, v0.20.0, v0.20.1) to confirm it's a regression
- [x] I have identified the root cause commit and analyzed the code change

### PR DESCRIPTION
**Pull request by liulanze**
## Purpose

Fix a regression introduced by #40860 (DeepSeek V4 support) where `csrc/layernorm_kernels.cu` was changed to always upcast the RMSNorm weight to `float32` before the final multiplication. The correct behaviour, per the Python spec in `vllm/ir/ops/layernorm.py:20`, is to multiply in the weight's native dtype (BF16/FP16):

```python
x = x.to(weight.dtype) * weight  # multiply in weight.dtype
```

The FP32 weight path introduces ~0.03 per-element error per layer. For models with many Q/K RMSNorm layers (e.g. LLaDA-2, DeepSeek with Q/K norm) this compounds catastrophically — 20-layer cumulative divergence exceeds 100, and only ~2% of output tokens match the reference.

The bug was reproduced locally on an A100-SXM4-40GB before implementing the fix, confirming the reported max diff of ~3e-02. The fix was built from source and verified on the same hardware.

Fixes #42325

## Changes

**`csrc/layernorm_kernels.cu`** — 3 locations:
- `rms_norm_kernel` vectorized path
- `fused_add_rms_norm_kernel` FP16/BF16 optimised path
- `fused_add_rms_norm_kernel` generic scalar fallback

Pattern in each location:
```cpp
// Before (regressed in #40860):
float w = static_cast<float>(weight_elem);
out = static_cast<scalar_t>(x * s_variance * w);

// After (restores v0.19.1 behaviour):
out = static_cast<scalar_t>(x * s_variance) * weight_elem;
```

**`csrc/layernorm_quant_kernels.cu`** — 3 matching locations:
The FP8 quant kernels contain an explicit comment explaining they were designed to match the non-fused composite path (`rms_norm` output → `static_scaled_fp8_quant`). After fixing the non-fused path, the quant kernels require the same update to stay consistent. Without it, `test_fused_rms_norm_quant` fails because the two paths diverge.

## Test Plan

```bash
pytest tests/kernels/core/test_layernorm.py -x -q
pytest tests/kernels/ir/test_layernorm.py -x -q
```

Regression repro script (shows `FIXED` after patch):

```python
import torch
from vllm import _custom_ops as ops

x = torch.randn(48, 128, dtype=torch.bfloat16, device='cuda')
weight = torch.randn(128, dtype=torch.bfloat16, device='cuda')
out = torch.empty_like(x)
ops.rms_norm(out, x, weight, 1e-6)

x_fp32 = x.to(torch.float32)
variance = x_fp32.pow(2).mean(-1, keepdim=True)
x_normed = x_fp32 * torch.rsqrt(variance + 1e-6)
ref = (x_normed.to(weight.dtype) * weight).to(x.dtype)

diff = (out.float() - ref.float()).abs().max().item()
print(f"Max diff: {diff:.6e}")
print("FIXED" if diff < 0.001 else "BUG CONFIRMED")
```

## Test Result

Reproduced the bug and verified the fix locally on an A100-SXM4-40GB, building vllm from source on this branch:

| | Before (bug reproduced) | After (fix verified) |
|---|---|---|
| `rms_norm` max diff (BF16) | `3.125e-02` ✗ | `0.000e+00` ✓ |
| `fused_add_rms_norm` max diff (BF16) | `3.125e-02` ✗ | `0.000e+00` ✓ |

```
tests/kernels/core/test_layernorm.py: 865 passed
tests/kernels/ir/test_layernorm.py:   1442 passed, 361 skipped
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

Note summary:

DeepSeek V4 support accidentally changed 6 CUDA kernel sites to upcast weights to float32 before multiplying, deviating from the spec (multiply in weight's native BF16 dtype); tiny per-layer rounding errors compounded across 20+ norm layers until the model output was nearly completely wrong — fixed by restoring the correct cast-then-multiply order.

### CODE REVIEW

**Comment by liulanze** [comment]
Hi @AlonKellner-RedHat would appreciate your review when you get a chance. Happy to discuss the approach if anything is unclear. Thanks.

**Comment by liulanze** [comment]
Hi @yewentao256 / @mgoin, this touches layernorm_kernels.cu which has no CODEOWNER. Would appreciate a review when you have a moment!

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work!
Please add a `lm_eval` metrics to show we don't hurt in e2e acc

**Comment by liulanze** [comment]
> Thanks for the work! Please add a `lm_eval` metrics to show we don't hurt in e2e acc

Thanks @yewentao256, here are the `lm_eval` results comparing `upstream/main` (v0.21.0, unfixed kernels) against this fix branch, both using `TinyLlama/TinyLlama-1.1B-Chat-v1.0` in BF16 on an H100 80GB:

| Task          | Metric   | Baseline (v0.21.0) | Fix branch         |
|---------------|----------|--------------------|--------------------|
| arc_challenge | acc      | 0.3046 ± 0.0134    | 0.3046 ± 0.0134    |
| arc_challenge | acc_norm | 0.3276 ± 0.0137    | 0.3276 ± 0.0137    |
| hellaswag     | acc      | 0.4652 ± 0.0050    | 0.4652 ± 0.0050    |
| hellaswag     | acc_norm | 0.6044 ± 0.0049    | 0.6044 ± 0.0049    |

Scores are identical with no regression. `TinyLlama` (RMSNorm, BF16) hits the fixed kernels, and I built from source to ensure the patched files were used.

**Review by yewentao256 — COMMENTED** [review]
Please solve the conflicts

**Comment by liulanze** [comment]
> Please solve the conflicts

Rebased onto latest main, the kernels have since moved to `csrc/libtorch_stable/`. Could you take another look? Thanks!

**Review by yewentao256 — APPROVED** [review]
LGTM, thanks for the work!

**Comment by liulanze** [comment]
@yewentao256 CI checks failing on this PR are unrelated to the change.
`kernels-fusedmoe-layer-test-2-b200s` died on NVSHMEM/IBRC connection setup, and both MI300 multimodal jobs hit a simultaneous "GPU Hang" hardware event on node-2. 
Could you please help to proceed with merge? Thanks!

**Comment by AndreasKaratzas** [comment]
I don't see why this PR would create a problem but given that the proposed changes are CU level and that one of the tests is failing with a seg fault, I would advise to investigate this a bit further. We are also transposing to MI325 cluster that should be more stable. 

cc @yewentao256

**Review by yewentao256 — COMMENTED** [review]
CC @liulanze for the seg fault

**Comment by AndreasKaratzas** [comment]
Also we just migrated to different cluster, so probably rebase too so that it runs on mi325 instead of mi300.

**Comment by AndreasKaratzas** [comment]
Things look far better now, except for the Lora test, but in the logs it is a file not found, so I am thinking it is unrelated to this PR. Let's wait for the rest of the tests and then merge.

**Comment by AndreasKaratzas** [comment]
Let me also confirm the multi-modal failure on AMD. Looks unrelated but give me 2 hrs to confirm if possible.

**Comment by AndreasKaratzas** [comment]
The multi modal test on AMD is not a regression of this PR, this PR can safely be force merged, I am going to take a look into the AMD failure in a different PR.

**Review by AndreasKaratzas — APPROVED** [review]

**Comment by liulanze** [comment]
@AndreasKaratzas Thanks for the review!

**Comment by liulanze** [comment]
Friendly ping, this PR is approved by @yewentao256 and @AndreasKaratzas, and `lm_eval` on `TinyLlama-1.1B/H100` shows identical scores. The remaining CI failures are unrelated to this change.

@mgoin @WoosukKwon - would you be able to force-merge when you get a chance? Happy to address anything else if needed. Thanks!

**Comment by AndreasKaratzas** [comment]
I posted the PR in slack to get it force merged.

**Comment by mgoin** [comment]
I'm not sure why accuracy on tiny llama is taken as a proxy here for deepseekv4, as that is the model that changed the behavior.. maybe @zyongye or @ivanium has more context possibly on the change

**Comment by zyongye** [comment]
Only check with python IR implementation is not ideal as it is written follow the CUDA implementation. And ideally all non Tensor core operation shall be operate on FP32 to maximize precision. If the score has no regression, then there's no point change the behavior? And I think we need to change the IR side instead of CUDA impl.
